marketplacetests/marketplace/app.py

```diff
@@ -64,12 +64,30 @@ def offline_message_text(self):
         self.wait_for_element_displayed(*self._offline_message_locator)
         return self.marionette.find_element(*self._offline_message_locator).text
 
+    @property
+    def notification_message_text(self):
+        return self.marionette.execute_script('return window.document.getElementById("notification-content").innerHTML;')
+
+    def check_notification_message_for_text(self, text):
+        current_text = self.notification_message_text
+        # TODO: Added logging to help debug common intermittent failures
+        # This can be removed in the future if the intermittents disappear
+        print 'Looking for: "%s" in notification. Found: "%s"' % (text, current_text)
+        return text in current_text
+
     def wait_for_notification_message(self, message):
+        """This will wait for the specified message to appear in the DOM element
+           for the notification message, not for that message to be visible.
+
+           This is required as often the message is no longer visible when we check,
+           but the expected text still exists in the DOM element.
+
+           This is also the reason that we wait for the element to no longer be
+           visible at the end of this method, whereas we do not wait for it to
+           first be visible.
+        """
+        Wait(self.marionette).until(lambda m: self.check_notification_message_for_text(message))
```

Comment:
`until` takes a callable object so there's no need for a lambda. The marionette object is passed when the object is called. Here's an untested example:

```python
def wait_for_notification_message(self, message):
    Wait(self.marionette).until(self._notification_exists(self, message))

class _notification_exists(object):

    def __init__(self, app, message):
        self.app = app
        self.expected = message

    def __call__(self, marionette):
        notification = marionette.find_element(*self.app._notification_locator)
        actual = marionette.execute_script('return arguments[0].innerHTML;', [notification])
        marionette.log('Looking for "%s" in "%s"' % (self.expected, actual))
        return self.expected in actual
```